File: android/spfn-core/src/main/kotlin/xyz/superfunction/spfn/core/SpfnErrorEnvelope.kt
// SPFN Mobile — the error envelope every SPFN endpoint answers with.
//
// Counterpart of Sources/SPFNCore/SPFNErrorEnvelope.swift. Hand-written and stable, like
// the operation descriptor beside it: the envelope is the one response shape that is not
// per-operation, so nothing generates it. It sits in core rather than in the client
// because generated decoders read it, and the hand-written `toString` stays in the same
// file as the type because every field is text a server chose — separating the two would
// leave a class one `data` keyword away from printing a payload into every stack trace.

package xyz.superfunction.spfn.core

/**
 * The canonical error envelope every SPFN endpoint answers with.
 *
 * Every field is text a server chose. A server can put anything in [message] or
 * [requestId] — including a session identifier it echoed back — so none of them may
 * reach a log by default.
 *
 * Deliberately not a `data class`: the generated `toString` would print all three, and
 * a `Throwable` carrying one prints its message into every stack trace. `equals` and
 * `hashCode` are written out by hand so nothing else changes, and the fields stay
 * ordinary public properties, so classifying an error is unaffected.
 */
class SpfnErrorEnvelope(
    val code: String,
    val message: String,
    val requestId: String
)
{
    override fun equals(other: Any?): Boolean =
        other is SpfnErrorEnvelope &&
            other.code == code &&
            other.message == message &&
            other.requestId == requestId

    override fun hashCode(): Int = (31 * (31 * code.hashCode() + message.hashCode())) + requestId.hashCode()

    override fun toString(): String = "SpfnErrorEnvelope(code=redacted, message=redacted, requestId=redacted)"

    /** The canonical form of this envelope, so a client can assert on exact bytes. */
    fun canonicalValue(): SpfnCanonicalValue = SpfnCanonicalValue.Obj(
        mapOf(
            "error" to SpfnCanonicalValue.Obj(
                mapOf(
                    "code" to SpfnCanonicalValue.Text(code),
                    "message" to SpfnCanonicalValue.Text(message),
                    "requestId" to SpfnCanonicalValue.Text(requestId)
                )
            )
        )
    )

    companion object
    {
        /**
         * Reads the envelope out of a parsed response body.
         *
         * An unrecognised code is not mapped onto a neighbouring one — that is the
         * generated `SpfnGeneratedErrorCode`'s job, and it rejects instead of guessing.
         */
        fun decode(value: SpfnCanonicalValue): SpfnErrorEnvelope
        {
            val root = SpfnDecoding.obj(value, "\$");
            val error = SpfnDecoding.obj(root["error"] ?: SpfnCanonicalValue.Null, "\$.error");
            return SpfnErrorEnvelope(
                code = SpfnDecoding.string(error["code"], "\$.error.code"),
                message = SpfnDecoding.string(error["message"], "\$.error.message"),
                requestId = SpfnDecoding.string(error["requestId"], "\$.error.requestId")
            );
        }
    }
}

File: Sources/SPFNCore/SPFNErrorEnvelope.swift
// SPFN Mobile — the error envelope every SPFN endpoint answers with.
//
// Hand-written and stable, like the operation descriptor beside it: the envelope is the
// one response shape that is not per-operation, so nothing generates it. It sits in core
// rather than in the client because generated decoders read it, and the redaction stays
// in the same file as the type because every field is text a server chose — separating
// the two would leave a plain struct whose `description` prints a payload.

/// The canonical error envelope every SPFN endpoint answers with.
///
/// Every field is text a server chose. A server can put anything in `message` or
/// `requestId` — including a session identifier it echoed back — so none of them may
/// reach a log by default. The redaction at the bottom of this file is what makes that
/// true; a caller that wants a field reads the property and decides for itself.
public struct SPFNErrorEnvelope: Equatable, Sendable
{
    public let code: String
    public let message: String
    public let requestID: String

    public init(code: String, message: String, requestID: String)
    {
        self.code = code
        self.message = message
        self.requestID = requestID
    }

    /// Reads the envelope out of a parsed response body.
    ///
    /// An unrecognised code is not mapped onto a neighbouring one — that is the
    /// generated `SPFNGeneratedErrorCode`'s job, and it rejects instead of guessing.
    public static func decode(_ value: SPFNCanonicalValue) throws -> SPFNErrorEnvelope
    {
        let root = try SPFNDecoding.object(value, at: "$")
        let error = try SPFNDecoding.object(
            root["error"] ?? .null,
            at: "$.error"
        )
        return SPFNErrorEnvelope(
            code: try SPFNDecoding.string(error["code"], at: "$.error.code"),
            message: try SPFNDecoding.string(error["message"], at: "$.error.message"),
            requestID: try SPFNDecoding.string(error["requestId"], at: "$.error.requestId")
        )
    }

    /// The canonical form of this envelope, so a client can assert on exact bytes.
    public var canonicalValue: SPFNCanonicalValue
    {
        .object([
            "error": .object([
                "code": .string(code),
                "message": .string(message),
                "requestId": .string(requestID),
            ]),
        ])
    }
}

// The default description of a struct prints every stored property, and `dump` and
// `String(reflecting:)` reach the same values through the synthesized mirror even when
// only `description` is overridden. All three doors are closed here rather than one:
// the payload is server-controlled text, and closing one door would just move the leak.
//
// `code`, `message` and `requestID` stay ordinary public properties, so classifying an
// error is unaffected. Only printing one by accident is.
extension SPFNErrorEnvelope: CustomStringConvertible, CustomDebugStringConvertible, CustomReflectable
{
    public var description: String
    {
        "SPFNErrorEnvelope(code: redacted, message: redacted, requestID: redacted)"
    }

    public var debugDescription: String
    {
        description
    }

    public var customMirror: Mirror
    {
        Mirror(self, unlabeledChildren: [Any]())
    }
}
